tendermint/src/chain/id.rs

```diff
@@ -133,38 +147,81 @@ impl<'de> Deserialize<'de> for Id {
     }
 }
 
+/// Validate that a given input string is a well-formed chain ID.
+const fn validate(s: &str) -> Result<(), ()> {
+    let bytes = s.as_bytes();
+    let mut i = 0;
+
+    while i < bytes.len() {
+        if !matches!(bytes[i], b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'-' | b'_' | b'.') {
+            return Err(());
+        }
+
+        i += 1;
+    }
+
+    Ok(())
+}
+
 #[cfg(test)]
 mod tests {
     use super::*;
     use crate::error::ErrorDetail;
 
     const EXAMPLE_CHAIN_ID: &str = "gaia-9000";
+    const MAX_LENGTH_CHAIN_ID: &str = "01234567890123456789012345678901234567890123456789";
+    const OVERLENGTH_CHAIN_ID: &str = "012345678901234567890123456789012345678901234567890";
```

Comment:
These need to be constants to test `Id::new`, which accepts a `&'static str` as an argument.